packages/runtime-core/src/compat/instance.ts

```diff
@@ -43,6 +43,7 @@ export type LegacyPublicInstance = ComponentPublicInstance &
   LegacyPublicProperties
 
 export interface LegacyPublicProperties {
+  $root: LegacyPublicInstance
```

Comment:
@philippevezina can you provide an example where this is needed? 

`$root` is already provided by `ComponentPublicInstance` so is implicitly available on `LegacyPublicInstance`

Reply:
Yes, `$root` is already included, but it's type is `ComponentPublicInstance`. It does not include the legacy Vue 2 properties included with `@vue/compat`.

This overrides the property's type, and includes properties such as `$on` to `$root`, which is available with `@vue/compat.`

Reply:
Here is an example of minimal reproduction: https://github.com/philippevezina/compat-root-reproduction/blob/main/src/App.vue#L19

You can clone the project to see the Typescript error. In this case, `$on` exists on `$root` with `@vue/compat` but displays incorrect typing. This is also the case for all other properties of `LegacyPublicProperties` which are also available in `$root`.

Reply:
I'm not well verse on Vue compat, but not seeing how adding `$root` would solve the issue, the example you sent me, the types are for `DefineComponent` and `LegacyPublicProperties` is not referenced, please correct me if I'm wrong.

Reply:
I'm not really familiar with how the types are supposed to work with Vue compat, but from experimenting with the reproduction I don't understand how the proposed change would help.

Running `vue-tsc` on that project does fail for `this.$root?.$on`, but it also fails if I try to use `this.$set`. This suggests it isn't using the legacy types for the `this` value.

The legacy types do seem to be used for the return value of `new Vue()`. I suppose this change would improve the type for `$root` in that scenario, though it's not clear to me whether that has much practical value.

If `$root` is being changed, I wonder whether the type for `$parent` should be changed too?

I also notice that the current types include `$children: LegacyPublicProperties[]`. It isn't clear to me why that is using `LegacyPublicProperties` rather than `LegacyPublicInstance`.